streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java

```diff
@@ -668,4 +693,195 @@ boolean sendingOldValueEnabled() {
+        // 1) Loads the data into the new state store. This data is accessed by the KTableKTablePrefixJoin processor.
+        // 2) Gets the foreignKey value data from the Other KTable and creates a SubscriptionWrapperResponse.
+        //    Returns the data keyed on K. Discards the CombinedKey as it is no longer needed after this stage.
+        final String thisStateStoreName = builder.newStoreName(STATE_STORE_NAME + "-thisSSName-");
+        final ForeignKeySingleLookupProcessorSupplier<K, KO, VO> oneToOne =
+                new ForeignKeySingleLookupProcessorSupplier<>(thisStateStoreName, ((KTableImpl<KO, VO, VO>) other).valueGetterSupplier());
+
+        final RocksDbKeyValueBytesStoreSupplier thisRocksDBRef = new RocksDbKeyValueBytesStoreSupplier(thisStateStoreName, true);
+
+        final StoreBuilder<KeyValueStore<Bytes, byte[]>> prefixScanStoreBuilder = Stores.timestampedKeyValueStoreBuilder(thisRocksDBRef,
+                combinedKeySerde,
+                new SubscriptionWrapperSerde());
+
+        //Performs other-table/foreign-key-driven updates
+        final ProcessorSupplier<KO, Change<VO>> prefixScanProcessorSupplier =
+                new KTableKTablePrefixScanProcessorSupplier<>(oneToOne.valueGetterSupplier());
+
+        //Need to write all updates to a given K back to the same partition, as at this point in the topology
+        //everything is partitioned on KO from the previous repartition step.
+        final String finalRepartitionerName = builder.newProcessorName(REPARTITION_NAME);
+        final String finalRepartitionTopicName = JOINOTHER_NAME + finalRepartitionerName;
+        final String finalRepartitionSourceName = builder.newProcessorName(SOURCE_NAME + finalRepartitionerName);
+        final String finalRepartitionSinkName = builder.newProcessorName(SINK_NAME + finalRepartitionerName);
+
+        //Create the processor to resolve the subscription updates.
+        final KTableValueGetterSupplier<K, V> thisKTableValueGetterSupplier = this.valueGetterSupplier();
+
+        final SubscriptionResolverJoinProcessorSupplier<K, V, VO, VR> resolverProcessor = new SubscriptionResolverJoinProcessorSupplier<>(thisKTableValueGetterSupplier, valueSerde().serializer(), joiner);
+        final String resolverProcessorName = builder.newProcessorName(KTableImpl.SOURCE_NAME + "resolver");
+
+        final KTableSource<K, VR> outputProcessor = new KTableSource<>(materializedInternal.storeName(), materializedInternal.queryableStoreName());
+        final String outputProcessorName = builder.newProcessorName(SOURCE_NAME + "output");
+
+        final HashSet<String> copartitions = new HashSet<>();
+        copartitions.add(repartitionSourceName);
+        copartitions.addAll(sourceNodes);
+        copartitions.addAll(((KTableImpl<?, ?, ?>) other).sourceNodes);
+        //TODO - Figure out a correct copartitioning strategy for topics with differing partition counts.
```

Comment:
Sorry, my comment was not very clear. Currently what you described DOES work for a single leftTable.joinOnForeignKey(rightTable, ...). However, I was concerned how it would work when we start chaining them together. I currently have a (somewhat complexly written) integration test (KTableKTableForeignKeyInnerJoinMultiIntegrationTest) where I do:

```
        CLUSTER.createTopic(TABLE_1, TABLE_1_SIZE, 1);
        CLUSTER.createTopic(TABLE_2, TABLE_2_SIZE, 1);
        CLUSTER.createTopic(TABLE_3, TABLE_3_SIZE, 1);
```
and then:
`TABLE_1.joinOnForeignKey(TABLE_2, ... ).joinOnForeignKey(TABLE_3, ...)`
The test does not pass (times-out after 60s of waiting) when `TABLE_1_SIZE != TABLE_3_SIZE` , using the co-partitioning as you suggested above. I'm not familiar enough with what's going on under the hood to explain it, nor have I spent much time digging deeper, but it seems that the partition distribution for chained `joinOnForeignKey` needs some attention.

Reply:
Ok, just for the record, I think we have to fix this before merging... I'll see if I can figure it out.

Reply:
I had researched it more extensively a few weeks ago  but just realized now I forgot to post the findings. They were pending publishing in my own review...

Reply:
I've just posted the findings. I am concerned about scope creep if we tinker with the copartitioner. As it stands right now I think the size of this PR is intimidating enough.